src/Control/Effect/Interpret.hs

```diff
@@ -53,27 +107,18 @@ instance (HFunctor eff, Carrier sig m) => Carrier (eff :+: sig) (InterpretC eff
 --   At time of writing, a simple use of 'runInterpretState' to handle a 'State' effect is about four times slower than using 'StateC' directly.
 --
 --   prop> run (runInterpretState (\ s op -> case op of { Get k -> runState s (k s) ; Put s' k -> runState s' k }) a get) === a
-runInterpretState :: (forall x . s -> eff (StateC s m) x -> m (s, x)) -> s -> InterpretStateC eff s m a -> m (s, a)
-runInterpretState handler state = runState state . runReader (HandlerState (\ eff -> StateC (\ s -> handler s eff))) . runInterpretStateC
-
-newtype InterpretStateC eff s m a = InterpretStateC { runInterpretStateC :: ReaderC (HandlerState eff s m) (StateC s m) a }
-  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadFix, MonadIO, MonadPlus)
-
-instance MonadTrans (InterpretStateC eff s) where
-  lift = InterpretStateC . lift . lift
-  {-# INLINE lift #-}
-
-newtype HandlerState eff s m = HandlerState (forall x . eff (StateC s m) x -> StateC s m x)
-
-runHandlerState :: (HFunctor eff, Functor m) => HandlerState eff s m -> eff (InterpretStateC eff s m) a -> StateC s m a
-runHandlerState h@(HandlerState handler) = handler . hmap (runReader h . runInterpretStateC)
+runInterpretState
+  :: ( HFunctor eff, Monad m )
+  => ( forall x . s -> eff ( StateC s m ) x -> m ( s, x ) )
+  -> s
+  -> ( forall t. Reifies t ( Handler eff ( StateC s m ) ) => InterpretC t eff ( StateC s m ) a )
+  -> m ( s, a )
+runInterpretState handler state m =
+  runState state $
+  runInterpret
+    ( \e -> StateC ( \s -> handler s e ) )
+    m
```

Comment:
This is such a cool demonstration of the effectiveness of this approach. Nicely done!

Reply:
:stars: